README.md
# AvatarLetter

It is a widget that uses the first few letters of a word to create an avatar with a set of basic functions.

## Preview

<img src="/al.gif" width="240" height="480"> | 
<img src="/demo_1.png" width="240" height="480"> | 
<img src="/demo_3.png" width="240" height="480"> |

## Getting Started

Add the plugin:

```yaml
dependencies:
  ...
  avatar_letter: ^1.0.1
```

# Basic Usage

```dart
    AvatarLetter(
                  size: 200,
                  backgroundColor: Colors.amber,
                  textColor: Colors.black,
                  fontSize: 100,
                  upperCase: true,
                  numberLetters: 1,
                  letterType: LetterType.Circular,
                  text: 'Test 01',
                ),
   ```
   
   ## Customization (Optional)

### AvatarLetter
**letterType** - Sets the format of the avatar that can be: (Rectangle, circular, none). Default to Rectangle<br/>
**text** - The text that will be used to extract the letters. Default to '?'<br/>
**textColor** - Sets the color of the letters. Default to Black<br/>
**textColorHex** - Sets the color of the letters from a valid hexadecimal code. Default to #ffffff<br/>
**backgroundColor** - Represents the background color of the avatar. Default to Black <br/>
**backgroundColorHex** -  Represents the background color of the avatar from a valid hexadecimal code. Default to #000000<br/>
**numberLetters** - Sets the number of letters to be extracted from the text. Default to 1 <br/>
**fontWeight** - Sets the font style. Default to Bold  <br/>
**fontFamily** - Sets the font .
**fontSize** - Sets the size of the letters within the avatar. Default to 16<br/>
**upperCase** - Used to set the uppercase letters of the avatar. Default to False <br/>
**size** - Sets the size of the avatar. Default to 50 <br/>


CHANGELOG.md
## [1.0.1] - Avatar letter update.
## - Improvements 

## [1.0.0] - Stable update.

## [0.1.1] - Few updates.

## [0.1.0] - Initial beta release.

* Beta release.


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:avatar_letter/avatar_letter.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/avatar_letter.dart
library avatar_letter;

import 'package:flutter/material.dart';

enum LetterType { Rectangle, Circular, None }

Color parseColor({String hexCode}) {
  String hex = hexCode.replaceAll("#", "");
  if (hex.isEmpty) hex = "000000";
  if (hex.length == 3) {
    hex =
        '${hex.substring(0, 1)}${hex.substring(0, 1)}${hex.substring(1, 2)}${hex.substring(1, 2)}${hex.substring(2, 3)}${hex.substring(2, 3)}';
  }
  Color col = Color(int.parse(hex, radix: 16));
  return col;
}

class AvatarLetter extends StatelessWidget {
  LetterType letterType;
  final String text;
  final double fontSize;
  final FontWeight fontWeight;
  final String fontFamily;
  Color backgroundColor;
  String backgroundColorHex;
  Color textColor;
  String textColorHex;
  double size;
  final int numberLetters;
  final bool upperCase;

  AvatarLetter(
      {Key key,
      this.letterType,
      @required this.text,
      @required this.textColor,
      @required this.textColorHex,
      @required this.backgroundColor,
      @required this.backgroundColorHex,
      this.size,
      this.numberLetters = 1,
      this.fontWeight = FontWeight.bold,
      this.fontFamily,
      this.fontSize = 16,
      this.upperCase = false}) {
    assert(numberLetters > 0);
  }

  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    letterType = (letterType == null) ? LetterType.Rectangle : letterType;
    size = (size == null || size < 30.0) ? 50.0 : size;
    backgroundColor = _colorBackgroundConfig();
    textColor = _colorTextConfig();
    return _leeterView();
  }

  Color _colorBackgroundConfig() {
    if (backgroundColor == null && backgroundColorHex == null)
      return Colors.black;
    else if (backgroundColor == null && backgroundColorHex != null)
      return parseColor(hexCode: backgroundColorHex);
    return backgroundColor;
  }

  Color _colorTextConfig() {
    if (textColor == null && textColorHex == null)
      return Colors.white;
    else if (textColor == null && textColorHex != null)
      return parseColor(hexCode: textColorHex);
    return textColor;
  }

  String _runeString({String value}) {
    return String.fromCharCodes(value.runes.toList());
  }

  String _textConfig() {
    var newText = text == null ? '?' : _runeString(value: text);
    newText = upperCase ? newText.toUpperCase() : newText;
    var arrayLeeters = newText.trim().split(' ');
    if (arrayLeeters != null) {
      if (arrayLeeters.length > 1 && arrayLeeters.length == numberLetters) {
        return '${arrayLeeters[0][0].trim()}${arrayLeeters[1][0].trim()}';
      }
      return '${newText[0]}';
    }
  }

  Widget _buildText() {
    return Text(
      _textConfig(),
      style: TextStyle(
        color: textColor,
        fontSize: fontSize,
        fontWeight: fontWeight,
        fontFamily: fontFamily,
      ),
    );
  }

  _buildTypeLeeter() {
    switch (letterType) {
      case LetterType.Rectangle:
        return RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(5.0),
        );
      case LetterType.Circular:
        return RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(size / 2),
        );
      case LetterType.None:
        return RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(0.0),
        );
    }
  }

  Widget _leeterView() {
    return Container(
      child: Material(
        shape: _buildTypeLeeter(),
        color: backgroundColor,
        child: Container(
          height: size,
          width: size,
          child: Center(
            child: _buildText(),
          ),
        ),
      ),
    );
  }
}


demo/lib/main.dart
import 'package:avatar_letter/avatar_letter.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _input = "avatar letter";
  Color _colorBackgroundHex = Colors.blueGrey;
  Color _colorText = Colors.white;
  int _numberLetters = 1;
  LetterType _letterType = LetterType.None;

  String _selectLetterTypes = 'None';
  String _selectNumberTypes = '1';
  String _selectColorBackground = 'blueGrey';
  String _selectColorText = 'white';
  bool _upperCase = false;

  static const letterTypes = <String>[
    'Rectangle',
    'Circular',
    'None',
  ];

  static const numberLetters = <String>[
    '1',
    '2',
  ];

  static const colors = <String>[
    'blueGrey',
    'white12',
    'green',
    'white',
    'amber',
    'black',
  ];

  final List<DropdownMenuItem<String>> _letterMenu = letterTypes
      .map((String value) => DropdownMenuItem<String>(
            value: value,
            child: Text(value),
          ))
      .toList();
  final List<DropdownMenuItem<String>> _letterNumbersMenu = numberLetters
      .map((String value) => DropdownMenuItem<String>(
            value: value,
            child: Text(value),
          ))
      .toList();

  final List<DropdownMenuItem<String>> _letterColorBackgroundMenu = colors
      .map((String value) => DropdownMenuItem<String>(
            value: value,
            child: Text(value),
          ))
      .toList();

  final List<DropdownMenuItem<String>> _letterColorTextMenu = colors
      .map((String value) => DropdownMenuItem<String>(
            value: value,
            child: Text(value),
          ))
      .toList();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Container(
            margin: EdgeInsets.all(16.0),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                AvatarLetter(
                  size: 200,
                  backgroundColor: _colorBackgroundHex,
                  textColor: _colorText,
                  fontSize: 100,
                  upperCase: _upperCase,
                  numberLetters: _numberLetters,
                  letterType: _letterType,
                  text: _input,
                ),
                SizedBox(
                  height: 50,
                ),
                ListTile(
                  title: Text('UpperCase'),
                  trailing: Switch(value: _upperCase, onChanged: (bool value){
                    setState(() {
                      _upperCase = value;
                    });
                  }),
                ),
                ListTile(
                  title: Text('Letter Type'),
                  trailing: DropdownButton<String>(
                    value: _selectLetterTypes,
                    items: _letterMenu,
                    onChanged: (String value) {
                      _selectLetterTypes = value;
                      setState(() {
                        if (letterTypes[0] == value) {
                          _letterType = LetterType.Rectangle;
                        } else if (letterTypes[1] == value) {
                          _letterType = LetterType.Circular;
                        } else if (letterTypes[2] == value) {
                          _letterType = LetterType.None;
                        }
                      });
                    },
                  ),
                ),
                ListTile(
                  title: Text('Number letter '),
                  trailing: DropdownButton<String>(
                    value: _selectNumberTypes,
                    items: _letterNumbersMenu,
                    onChanged: (String value) {
                      _selectNumberTypes = value;
                      setState(() {
                        if (numberLetters[0] == value) {
                          _numberLetters = 1;
                        } else if (numberLetters[1] == value) {
                          _numberLetters = 2;
                        }
                      });
                    },
                  ),
                ),
                ListTile(
                  title: Text('Color background '),
                  trailing: DropdownButton<String>(
                    value: _selectColorBackground,
                    items: _letterColorBackgroundMenu,
                    onChanged: (String value) {
                      _selectColorBackground = value;
                      setState(() {
                        if (colors[0] == value) {
                          _colorBackgroundHex = Colors.blueGrey;
                        } else if (colors[1] == value) {
                          _colorBackgroundHex = Colors.white12;
                        } else if (colors[2] == value) {
                          _colorBackgroundHex = Colors.green[800];
                        } else if (colors[3] == value) {
                          _colorBackgroundHex = Colors.white;
                        } else if (colors[4] == value) {
                          _colorBackgroundHex = Colors.amber;
                        }else if (colors[5] == value) {
                          _colorBackgroundHex = Colors.black;
                        }
                      });
                    },
                  ),
                ),
                ListTile(
                  title: Text('Color text '),
                  trailing: DropdownButton<String>(
                    value: _selectColorText,
                    items: _letterColorTextMenu,
                    onChanged: (String value) {
                      _selectColorText = value;
                      setState(() {
                        if (colors[0] == value) {
                          _colorText = Colors.blueGrey;
                        } else if (colors[1] == value) {
                          _colorText = Colors.white12;
                        } else if (colors[2] == value) {
                          _colorText = Colors.green[800];
                        } else if (colors[3] == value) {
                          _colorText = Colors.white;
                        } else if (colors[4] == value) {
                          _colorText = Colors.amber;
                        }else if (colors[5] == value) {
                          _colorText = Colors.black;
                        }
                      });
                    },
                  ),
                ),
                Container(
                  margin: EdgeInsets.only(left: 16.0, right: 16.0),
                  child: TextField(
                    decoration: InputDecoration(
                      hintText: _input,
                    ),
                    onChanged: (value) {
                      setState(() {
                        print('${_input}');
                        _input = value.isEmpty ? '?' : value;
                      });
                    },
                  ),
                )
              ],
            )),
      ),
    );
  }
}
